src/gpuarray/buffer.h

```diff
@@ -757,6 +763,48 @@ typedef enum _ga_usefl {
   GA_USE_OPENCL =   0x4000,
 } ga_usefl;
 
+/************************************************************************************
+*                         multi-gpu communicator interface                         *
+************************************************************************************/
+
+/**
+ * \brief Creates a new communicator (multi process version)
+ * \param clique_id [const char*] TODO
+ * \param ndev [int] TODO
+ * \param rank [int] TODO
+ * \param res [int*] TODO
+ * \return gpucomm* TODO
+ */
+GPUARRAY_PUBLIC gpucomm* gpucomm_init(const char* clique_id, int ndev, int rank,
+                                      int* res);
+
+/**
+ * \brief Frees resources associated with communicator object.
+ * \param comm [gpucomm*] TODO
+ * \return void TODO
+ */
+GPUARRAY_PUBLIC void gpucomm_destroy(gpucomm* comm);
+
+/**
+ * \brief Returns nice error message.
+ * \param comm [gpucomm*] TODO
+ * \param err [int] TODO
+ * \return const char* TODO
+ */
+GPUARRAY_PUBLIC const char* gpucomm_error(gpucomm* comm, int err);
+
+// TODO cuda extension:
+// 1. ncclGetUniqueId
+// 2. ncclCommCount (below mby ops in gpucomm struct?!)
+// 3. ncclCommUserRank
+// 4. ncclCommCuDevice
```

Comment:
I see that you are right about the id.  I somehow had some other idea in mind about how it worked.

The problem I have with a cuda extension is that we still need a way to pass it through as the clique_id parameter.  This would make the extension a required part for it to work, which makes it not exactly an extension, but rather a different form of binding.

I would prefer some wrapper that returns a `void *` or `char *` pointer.  However the problem remains that we need to tell the user how big it is so that they can copy it over to the other processes and use it.  We could always mimic the approach that nccl has taken with a dummy struct of a given size or we could return a pointer and size.